text/0000-safer-transmute.md

```diff
@@ -326,63 +319,41 @@ pub mod crate_a {
-            assert!(len <= N && len > 0);
-            Self(
-                private::NonEmptySliceInner {
-                    data: arr as *const T,
-                    len,
-                    lifetime: core::marker::PhantomData,
-                }
-            )
-        }
+#### Well-Definedness Does Not Imply Stability
+Since the well-definedness of a transmutation is affected by the layouts of the source and destination types, internal changes to those types' layouts may cause code which previously compiled to produce errors. In other words, transmutation causes a type's layout to become part of that type's API for the purposes of SemVer stability.
 
-        pub fn first(&self) -> &'a T {
-            unsafe { &*self.0.data }
-        }
-    }
 
-    // introduce a private module to avoid `private_in_public` error (E0446):
-    pub(crate) mod private {
-        #[repr(C)]
-        pub struct NonEmptySliceInner<'a, T> {
-            pub data: *const T,
-            pub len: usize,
-            pub lifetime: core::marker::PhantomData<&'a ()>,
-        }
-    }
+#### Signaling Safety and Stability with `Muckable`
+To signal that your type may be safely and safely constructed via transmutation, implement the `Muckable` marker trait:
+```rust
+use mem::transmute::Muckable;
 
-}
+#[derive(Muckable)]
+#[repr(C)]
+pub struct Foo(pub u8, pub u16);
 ```
-In the above example, the definitions of both `NonEmptySlice` and its field `NonEmptySliceInner` are marked `pub`, and all fields of these types are marked `pub`. However, `NonEmptySlice` is *not* fully implicitly constructible outside of `crate_a`, because the module containing `NonEmptySliceInner` is not visibile outside of `crate_a`.
 
-#### Constructability and Transmutation
-Transmutation supplies a mechanism for constructing instances of a type *without* invoking its implicit constructor, nor any constructors defined by the type's author. In the previous examples, it would be *unsafe* to transmute `[usize; 2]` into `NonEmptySlice` outside `crate_a`, because subsequent *safe* use of that value (namely, calling `first`) would violate memory safety:
-```rust
-/* ⚠️ This example intentionally does not compile. */
-// [usize; 2] ⟶ NonEmptySlice
-let _: NonEmptySlice<'static, u8> = transmute!([0usize; 2]); // Compile Error: `NonEmptySlice<_, _>` is not constructible here.
-```
-If a field is private, then instantiating or modifying it via transmutation is not, generally speaking, safe.
+The `Muckable` marker trait signals that your type's fields may be safely initialized and modified to *any* value. If you would not be comfortable making your type's fields `pub`, you probably should not implement `Muckable` for your type. By implementing `Muckable`, you promise to treat *any* observable modification to your type's layout as a breaking change. (Unobservable changes, such as renaming a private field, are fine.)
+
+As a rule, the destination type of a transmutation must be `Muckable`.
```

Comment:
Why just destination? When I transmute, e.g., `Vec` to something, I rely on stability of `Vec`'s layout.

The "safe initialization" aspect of `Muckable` is only relevant for the target type, but the "stable layout" aspect is relevant for both source and target.

Reply:
Since I think intertwining safety and stability is a mistake, my only hesitation is that this change imposes an all-or-nothing stability regime in yet another place. Total stability isn't *always* relevant for the source type. If you are using `TransmuteFrom` to write bytes of `Src` onto a wire (but not read them), the total stability of `Src` doesn't necessarily matter. If you are using `TransmuteFrom` to assert a type-level invariant but not actually transmuting, the total stability of `Src` doesn't necessarily matter. Requiring an all-or-nothing stability regime on both the source and destination types for safety withholds safe transmutation from the use-cases where these requirements are too coarse.


However, now that safety and stability are intertwined and stability is mostly all-or-nothing, I agree it makes sense to require that `Src` is layout-stable, too.

Reply:
Right, this is all a consequence of `Muckable` saying two things at once.